src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker.cs

```diff
@@ -0,0 +1,496 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.MSBuild.Build;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MSBuild
+{
+    public partial class MSBuildProjectLoader
+    {
+        private partial class Worker
+        {
+            private readonly Workspace _workspace;
+            private readonly DiagnosticReporter _diagnosticReporter;
+            private readonly PathResolver _pathResolver;
+            private readonly ProjectFileLoaderRegistry _projectFileLoaderRegistry;
+            private readonly ProjectBuildManager _buildManager;
+            private readonly string _baseDirectory;
+
+            /// <summary>
+            /// An ordered list of paths to project files that should be loaded. In the case of a solution,
+            /// this is the list of project file paths in the solution.
+            /// </summary>
+            private readonly ImmutableArray<string> _requestedProjectPaths;
+
+            /// <summary>
+            /// Global MSBuild properties to set when loading projects.
+            /// </summary>
+            private readonly ImmutableDictionary<string, string> _globalProperties;
+
+            /// <summary>
+            /// Map of <see cref="ProjectId"/>s, project paths, and output file paths.
+            /// </summary>
+            private readonly ProjectMap _projectMap;
+
+            /// <summary>
+            /// Progress reporter.
+            /// </summary>
+            private readonly IProgress<ProjectLoadProgress> _progress;
+
+            /// <summary>
+            /// Provides options for how failures should be reported when loading requested project files.
+            /// </summary>
+            private readonly DiagnosticReportingOptions _requestedProjectOptions;
+
+            /// <summary>
+            /// Provides options for how failures should be reported when loading any discovered project files.
+            /// </summary>
+            private readonly DiagnosticReportingOptions _discoveredProjectOptions;
+
+            /// <summary>
+            /// When true, metadata is preferred for any project reference unless the referenced project is already loaded
+            /// because it was requested.
+            /// </summary>
+            private readonly bool _preferMetadataForReferencedProjects;
```

Comment:
Maybe call this _preferMetadataForReferencedProjectsOutsideInitialSet?

Reply:
Sure. I'll update the name.

Reply:
I'm going with `_preferMetadataForReferencesOfDiscoveredProjects`